Invariant cleanup in goto programs i*

hannes-steffenhagen-diffblue opened the pull request:
This PR is part of the invariant cleanup work, for the files in goto-programs starting with i.

This entails replacing asserts with cprover style invariants, using structured exception types and using `to_XXX_expr` style conversions instead of 'structure' asserts for expressions.

hannes-steffenhagen-diffblue committed:
Use structured exceptions in initialize_goto_model

hannes-steffenhagen-diffblue committed:
Replace throws with structured throws or invariant in interpreter

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 6ff309661b2902df6ae29212a79faee190997950).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84818543

xbauch commented on src/goto-programs/interpreter_class.h:
```diff
@@ -279,8 +280,9 @@ class interpretert:public messaget
+    // FIXME not sure if this can happen
     if(v.size()!=1)
-      throw "invalid boolean value";
+      throw interpreter_errort("invalid boolean value");
```
I wonder if this shouldn't be an invariant. The input `expr` is always the path condition of the interpreted program and I don't think the user is directly responsible for what the interpreter chooses to be a path condition.

hannes-steffenhagen-diffblue committed:
Add pointer_offset_exprt

hannes-steffenhagen-diffblue committed:
Use structured invariant in interpreter_class.h

hannes-steffenhagen-diffblue committed:
Invariant cleanup in interpreter_evaluate.cpp

This mostly entails using to_XXX_expr in place of using ad hoc asserts.
At the same time, some minor cleanup work is done to improve the
readability of the affected code

hannes-steffenhagen-diffblue commented on src/goto-programs/initialize_goto_model.cpp:
```diff
@@ -69,21 +74,17 @@ goto_modelt initialize_goto_model(
-        msg.error() << "failed to figure out type of file" << messaget::eom;
-        throw 0;
+        throw invalid_source_file_exceptiont(
+          "Failed to figure out type of file `" + filename + '\'');
```
This should arguably happen before the check whether the file can be opened, but that's a minor point and not really within the spirit of this PR

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 79efa6da57ffe76f9c7c41e3ba11475038b0127a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87615645
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

xbauch reviewed (approved):
It might be a little bit more readable if the variable renaming was in a separate commit. But other than that, it looks good to me.

martin-cs commented on src/goto-programs/interpreter.cpp:
```diff
@@ -69,12 +69,12 @@ void interpretert::initialize(bool init)
 
   if(main_it==goto_functions.function_map.end())
-    throw "main not found";
+    throw analysis_exceptiont("main not found");
```
Is this actually a structural invariant on goto-programs?

hannes-steffenhagen-diffblue replied:
I wasn't sure about this one, because you might have goto programs without an entry point (libraries mostly). I suppose it could be a precondition if we assume that a caller should've checked for an entry point before calling this.

martin-cs replied:
It is getting in to the heart of why we need to check that goto-programs are well formed and to clarify what "normal forms" exist for them.  If it has been linked and is a complete goto program then it should have an entry point which is a function in the map.  If it is only partially built (goto-cc -c for example) then it may not do.  It is not immediately clear to me whose fault it would be to try to interpret a n "object file" rather than an "executable".

martin-cs commented on src/goto-programs/interpreter.cpp:
```diff
@@ -69,12 +69,12 @@ void interpretert::initialize(bool init)
 
   if(!goto_function.body_available())
-    throw "main has no body";
+    throw analysis_exceptiont("main has no body");
```
As above.

martin-cs commented on src/goto-programs/interpreter.cpp:
```diff
@@ -292,7 +292,10 @@ void interpretert::step()
-      throw "RETURN without call"; // NOLINT(readability/throw)
+    {
+      throw analysis_exceptiont(
+        "RETURN without call"); // NOLINT(readability/throw)
```
I think this is an INVARIANT as if the interpreter is working as it should, then this should never happen.

hannes-steffenhagen-diffblue replied:
I was thinking this could happen in a nonsensical goto program that has returns in unexpected places, such as `__CPROVER_initialize`. Is that wrong?

martin-cs replied:
If I understand the code correctly (and it is possible I don't) then the INVARIANT is that the number of RETURNS executed should never exceed the number of functions started.  So this should always hold.

martin-cs commented on src/goto-programs/interpreter.cpp:
```diff
@@ -317,19 +320,23 @@ void interpretert::step()
     trace_step.type=goto_trace_stept::typet::SPAWN;
-    throw "START_THREAD not yet implemented"; // NOLINT(readability/throw)
+    throw analysis_exceptiont(
+      "START_THREAD not yet implemented"); // NOLINT(readability/throw)
```
UNIMPLEMENTED ?  Also below.

martin-cs commented on src/goto-programs/interpreter.cpp:
```diff
@@ -359,7 +366,7 @@ void interpretert::step()
     break;
   default:
-    throw "encountered instruction with undefined instruction type";
+    UNREACHABLE; // I'm assuming we'd catch such a thing before this point
```
Sounds like you're saying it is a consequence of the goto program being well-formed ;-)

martin-cs commented on src/goto-programs/interpreter.cpp:
```diff
@@ -724,15 +730,15 @@ void interpretert::assign(
 {
   if(!evaluate_boolean(pc->guard))
-    throw "assumption failed";
+    throw analysis_exceptiont("assumption failed");
```
Unrelated to this; I'm not sure this is the right way of handling this case.  I think it should just terminate correctly.

martin-cs reviewed (approved):
Thanks for the clean-up.  Comments are mostly opinions on what kind of error conditions these are.

hannes-steffenhagen-diffblue commented:
@martin-cs Yes, since I'm not super familiar with this code myself I was a bit unclear in many cases about whether INVARIANT or exception is the better 'mechanism' to use. I'll have another look over the cases you pointed out.

martin-cs commented:
@hannes-steffenhagen-diffblue : The test I tend to you is : if there are a set of user controllable flags and a non-corrupted / incorrect program that can cause it to happen, it is a potential user error and should be an exception.  If the only way it can happen is if there are one or more bugs in CPROVER, then it should be an invariant of some kind.

hannes-steffenhagen-diffblue commented:
@martin-cs I completely agree with that, it's just difficult to judge on a case by case basis sometimes.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 11af1960df2ebe24c5a1bc209d52ae10046e2ad0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87644100
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

hannes-steffenhagen-diffblue commented:
Heads up, I'm a bit uncomfortable with putting these changes in as the interpreter seems to be broken and untested atm.

See https://github.com/diffblue/cbmc/issues/3146

I'll extract the initialize_goto_model stuff as that seems to be unrelated

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: ce7f7926415760b3f28543b5811e1788e68cccc8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87649507

martin-cs commented:
@hannes-steffenhagen-diffblue : It is difficult to judge; this is why I try to phrase my comments on these as "IMHO it should be".

As regards the interpreter; it didn't use to be in a great state but I thought someone at Diffblue had taken over it's care and feeding.  I have a vague feeling that @dcattaruzza was asking about it at one point...

tautschnig commented on src/goto-programs/interpreter_class.h:
```diff
@@ -279,8 +280,9 @@ class interpretert:public messaget
+    // FIXME not sure if this can happen
     if(v.size()!=1)
-      throw "invalid boolean value";
+      throw analysis_exceptiont("invalid boolean value");
```
I would drop the FIXME and turn this into an `INVARIANT`. People will start complaining if it's not actually an invariant... :-)

tautschnig reviewed (approved):
Looks ok to me (with one comment below); but really this should be checked against TG, where I believe some interpreter tests exist.

hannes-steffenhagen-diffblue commented:
I extracted the `initialize_goto_model` bit because that is tested, short, and fairly trivial (no changing of conditions, just different exception types) https://github.com/diffblue/cbmc/pull/3190
